empirical_calibration/core.py

```diff
@@ -375,17 +375,26 @@ def zoom(left, right, weights):
 
 def dmatrix_from_formula(formula: str, df: pd.DataFrame) -> pd.DataFrame:
   """Generates dmatrix from formula and dataframe.
-  This is a wrapper around patsy's dmatrix function.
+
+  Builds a design matrix by applying patsy's dmatrix function one variable at a
+  time to ensure that reference categories of categorical variables are not
+  dropped.
   Args:
-    formula: Formula to be fed into patsy. No outcome variable allowed.
+    formula: patsy-style formula. No outcome variable allowed. Interactions
+      between categorical variables of the form dim1*dim2 should not be used
+      (use dim1:dim2 instead).
     df: Df containing the raw data.
   Returns:
     Design matrix.
   """
-  dmatrix = patsy.highlevel.dmatrix(formula, df, return_type="dataframe")
-  if "Intercept" in dmatrix.columns:
-    dmatrix.drop(columns="Intercept", inplace=True)
-  return dmatrix
+
+  # Parse the formula into a list of dimensions
+  dimensions = formula.replace(" ", "").replace("~-1+", "").replace(
+      "~0+", "").replace("~1+", "").replace("~", "").split("+")
+  return pd.concat([
+      patsy.highlevel.dmatrix(
+          "~ 0 + " + dimj, df, return_type="dataframe") for dimj in dimensions
```

Comment:
nit: do you want to use f-string here?
nit: I'd spell out dimension.

`f"~ 0 + {dimension}", df, return_type="dataframe") for dimension in dimensions`

Reply:
Done. Thanks for the suggestion.